### ISSUE TITLE
Add first party linting support i.e. `tofu lint`

### ISSUE THREAD

**Issue by dragonfleas** [opentofu/opentofu#2213]
### OpenTofu Version

```shell
N/A
```

### The problem in your OpenTofu project

No linting tools for OpenTofu.

### Attempted Solutions

- Explored many different tools online
- Checked `tflint` which has [indirectly stated they will not support opentofu](https://github.com/terraform-linters/tflint/issues/2037#issuecomment-2094738720)
- Checked for style and quality enforcement from SAST tools, if it exists it's minimal

### Proposal

I have two proposals:

The first of which is full first party linting support using the OpenTofu binary package and running something like `tofu lint`, this will likely be very complicated as the knowledge around specific language specifications is as one in Slack put it *underspecified*. The benefits of this would be a massively increased user experience and ease of adoption for those seeking quality and style enforcement either in editor, or optionally during their plans and applies. There are many opinions about style, primitive usages, built in tofu function use cases, that we could all likely benefit from having some sort of static analysis as a quality gate for those things.

The second alternative proposal is, we go back to the last commit of `tflint` before the dual license [LICENSE_BUSL](https://github.com/terraform-linters/tflint/blob/master/LICENSE-BUSL) was introduced and fork it, creating a fork labeled `tofulint` and take over the project to work on and add the relevant versioned changes that are required from 1.5 era up to the current, I don't know if this would be more of an involved effort or not, but in either of these proposals I'd gladly pledge at least 10 hours a week to put towards this effort.

**Comment by apparentlymart** [comment]
Hi @dragonfleas! Thanks for opening this issue.

In my experience the term "lint" is often used for a wide variety of different kinds of analyses -- basically _anything_ that's more subtle than just "is this valid at all?" -- so I think that along with voting for the general idea of linting it would also be helpful for folks to share what sorts of analyses they are particularly interested in.

---

I have generally been most interested in analyses that might help explain why some particular configuration is not behaving as the author expected. I previously made a list of possible candidates which I will reproduce here just as a starting point, although I am not meaning this to say that any of the following is definitely required for the feature request in this issue to be met.

- Using `depends_on` to specify an explicit dependency that OpenTofu can already discover via an implicit dependency. This is not itself a harmful thing to do, but it is something I commonly see folks try when they are new to OpenTofu and don't yet have a solid mental model of how its dependency analysis works, and so explaining that such a declaration does _absolutely nothing_ tends to free a frustrated person to seek other explanations rather than getting stuck chasing something that hasn't materially changed their situation.
- Conditional expressions where the predicate provably always returns true or always false, such as comparing two values of different types with either `==` or `!=`.
- Using literals of a particular type in situations that would always lead to an implicit conversion to some other type. For example, "1" + 2 could warn that the "1" ought to be written as 1. This is not problematic in itself, but this is a thing in the category of "eliminate odd stuff that is likely to distract from finding a real cause". (I would hesitate to extend this to non-literal operands because relying on implicit conversions for resource attributes and similar is pretty common and so would likely be very noisy without really explaining anything useful.)
- Using the impure functions (currently `timestamp`, `uuid`, and `bcrypt`) in resource arguments, which will typically cause a configuration to never converge because there will always be a planned update/replace for that resource. (Could we detect that indrectly too, so that e.g. referring to a local value which is itself non-converging would also generate the warning?)
- A heuristic to check if it looks like a heredoc template is trying to generate JSON? We might not be able to detect this reliably enough, but if we can it would be nice to suggest using jsonencode as a more robust answer, since we commonly see folks trying to generate JSON using string templates and struggling to get all of the punctuation right. The goal of catching what appears to be slightly-invalid JSON makes this heuristic a bit trickier than it would otherwise be.
- Recommend explicit type constraints for all input variables (even if they are just any), because that tends to help understand why something weird is going on when a caller writes an expression of the wrong type.
- Using `lookup` in its two-argument form, which has long been superseded by the explicit index syntax.
- Using `lookup` in its three-argument form, which (subjectively) tends to be more readable as a `try` call instead, because the argument then appears just as normal traversal syntax and is thus easier to read.
- Perhaps we can detect legacy usage with splat expressions combined with index expressions as was common in much older versions, like `aws_instance.example.*.id[0]`, and propose writing `aws_instance.example[0].id` instead to make the expression easier to understand when debugging.
- We could report when a local value is declared but not referred to anywhere in the same module. A local value doesn't contribute to the externally-visible behavior of a module, so there's no point in declaring one that isn't used somewhere else in the module.
- Warn if the file path argument to functions like file, `templatefile`, `filebase64`, etc refers to a file that doesn't exist and seems similar to a filename in a location that seems like it's intending to generate a file during the apply operation, such as in a local-exec provisioner for a resource that this object depends on? This one might be too fuzzy to detect reliably, but if we could catch it then we could give more explicit direction that these functions aren't suitable for that purpose, and suggest using similar provider functionality instead.
- If a quoted literal string contains something which appears to be a valid reference to elsewhere in the module (it's both syntactically correct and resolves to something in the current module), hint that the author probably wanted to write it not in quotes. For example, `"var.example"` would, if there's also a `variable "example"` block in the module, suggest writing plain `var.example` instead.
- Applying the splat operator `[*]` to a resource which has `for_each` set. Although it's technically okay to do that, because the splat operator can apply to values of any type, for `for_each` resources in particular it's much more likely to be an incorrect attempt to mimic selection of multiple attributes on a count resource. This lint rule could suggest wrapping the resource expression itself in `values(...)` in order to get what was probably the intended effect, like `values(aws_instance.example)[*].private_ip`.

There isn't really any coherent overall theme in these, except perhaps the general idea that these are insights I've often shared with people while helping them understand why their modules didn't behave in the way they'd expected and so they are perhaps best described as "debugging hints".

---

Another pretty broad question is to what extent a solution to this feature request lives in OpenTofu Core vs. in plugins.

I think quite a lot of the analyses folks enjoy from tools like `tflint` are in the platform-specific rulesets like [tflint-ruleset-aws](https://github.com/terraform-linters/tflint-ruleset-aws). It's not really appropriate for _those specific rules_ to be checked by OpenTofu Core because the corresponding provider plugins evolve independently of core.

It would be interesting if providers plugins themselves could ship with a set of lint checks that form a good general baseline, so that e.g. anyone using the AWS provider automatically gets some generally-applicable set of AWS-provider-specific rules. However, I also think `tflint`'s model of allowing plugins to be developed independently of specific providers is helpful, because that allows for more interesting situations like linting on the interactions between two different providers, or having an organization write their own linting rules for special considerations that are unique to their own environment.

**Comment by dragonfleas** [comment]
> Hi @dragonfleas! Thanks for opening this issue.
> 
> In my experience the term "lint" is often used for a wide variety of different kinds of analyses -- basically _anything_ that's more subtle than just "is this valid at all?" -- so I think that along with voting for the general idea of linting it would also be helpful for folks to share what sorts of analyses they are particularly interested in.
> 
> I have generally been most interested in analyses that might help explain why some particular configuration is not behaving as the author expected. I previously made a list of possible candidates which I will reproduce here just as a starting point, although I am not meaning this to say that any of the following is definitely required for the feature request in this issue to be met.
> 
> * Using `depends_on` to specify an explicit dependency that OpenTofu can already discover via an implicit dependency. This is not itself a harmful thing to do, but it is something I commonly see folks try when they are new to OpenTofu and don't yet have a solid mental model of how its dependency analysis works, and so explaining that such a declaration does _absolutely nothing_ tends to free a frustrated person to seek other explanations rather than getting stuck chasing something that hasn't materially changed their situation.
> * Conditional expressions where the predicate provably always returns true or always false, such as comparing two values of different types with either `==` or `!=`.
> * Using literals of a particular type in situations that would always lead to an implicit conversion to some other type. For example, "1" + 2 could warn that the "1" ought to be written as 1. This is not problematic in itself, but this is a thing in the category of "eliminate odd stuff that is likely to distract from finding a real cause". (I would hesitate to extend this to non-literal operands because relying on implicit conversions for resource attributes and similar is pretty common and so would likely be very noisy without really explaining anything useful.)
> * Using the impure functions (currently `timestamp`, `uuid`, and `bcrypt`) in resource arguments, which will typically cause a configuration to never converge because there will always be a planned update/replace for that resource. (Could we detect that indrectly too, so that e.g. referring to a local value which is itself non-converging would also generate the warning?)
> * A heuristic to check if it looks like a heredoc template is trying to generate JSON? We might not be able to detect this reliably enough, but if we can it would be nice to suggest using jsonencode as a more robust answer, since we commonly see folks trying to generate JSON using string templates and struggling to get all of the punctuation right. The goal of catching what appears to be slightly-invalid JSON makes this heuristic a bit trickier than it would otherwise be.
> * Recommend explicit type constraints for all input variables (even if they are just any), because that tends to help understand why something weird is going on when a caller writes an expression of the wrong type.
> * Using `lookup` in its two-argument form, which has long been superseded by the explicit index syntax.
> * Using `lookup` in its three-argument form, which (subjectively) tends to be more readable as a `try` call instead, because the argument then appears just as normal traversal syntax and is thus easier to read.
> * Perhaps we can detect legacy usage with splat expressions combined with index expressions as was common in much older versions, like `aws_instance.example.*.id[0]`, and propose writing `aws_instance.example[0].id` instead to make the expression easier to understand when debugging.
> * We could report when a local value is declared but not referred to anywhere in the same module. A local value doesn't contribute to the externally-visible behavior of a module, so there's no point in declaring one that isn't used somewhere else in the module.
> * Warn if the file path argument to functions like file, `templatefile`, `filebase64`, etc refers to a file that doesn't exist and seems similar to a filename in a location that seems like it's intending to generate a file during the apply operation, such as in a local-exec provisioner for a resource that this object depends on? This one might be too fuzzy to detect reliably, but if we could catch it then we could give more explicit direction that these functions aren't suitable for that purpose, and suggest using similar provider functionality instead.
> * If a quoted literal string contains something which appears to be a valid reference to elsewhere in the module (it's both syntactically correct and resolves to something in the current module), hint that the author probably wanted to write it not in quotes. For example, `"var.example"` would, if there's also a `variable "example"` block in the module, suggest writing plain `var.example` instead.
> * Applying the splat operator `[*]` to a resource which has `for_each` set. Although it's technically okay to do that, because the splat operator can apply to values of any type, for `for_each` resources in particular it's much more likely to be an incorrect attempt to mimic selection of multiple attributes on a count resource. This lint rule could suggest wrapping the resource expression itself in `values(...)` in order to get what was probably the intended effect, like `values(aws_instance.example)[*].private_ip`.
> 
> There isn't really any coherent overall theme in these, except perhaps the general idea that these are insights I've often shared with people while helping them understand why their modules didn't behave in the way they'd expected and so they are perhaps best described as "debugging hints".
> 
> Another pretty broad question is to what extent a solution to this feature request lives in OpenTofu Core vs. in plugins.
> 
> I think quite a lot of the analyses folks enjoy from tools like `tflint` are in the platform-specific rulesets like [tflint-ruleset-aws](https://github.com/terraform-linters/tflint-ruleset-aws). It's not really appropriate for _those specific rules_ to be checked by OpenTofu Core because the corresponding provider plugins evolve independently of core.
> 
> It would be interesting if providers plugins themselves could ship with a set of lint checks that form a good general baseline, so that e.g. anyone using the AWS provider automatically gets some generally-applicable set of AWS-provider-specific rules. However, I also think `tflint`'s model of allowing plugins to be developed independently of specific providers is helpful, because that allows for more interesting situations like linting on the interactions between two different providers, or having an organization write their own linting rules for special considerations that are unique to their own environment.

All of those are really great recommendations, so many common code smells I see in new codebases (and frankly sometimes ones that I myself forget about in OpenTofu). I think I'd like to draw a parallel to how the Deno team has decided to handle first-party linting, as I think it offers a useful model for OpenTofu to consider.

Deno ships with built-in linting capabilities that focus on common pitfalls and best practices documented in their official guides, while remaining extensible for additional rules. They decided to do this because of the lack of tooling in the original ecosystem they were working to improve upon, and eventually replace. I believe OpenTofu could benefit from a similar approach, where core linting focuses on patterns that are:

1. Documented in official OpenTofu guides
2. Known to cause confusion or maintainability issues
3. Related to core language features rather than provider-specific concerns

Some specific suggestions building on your list:

- The `terraform_remote_state` data source usage could be flagged with suggestions to use more modern patterns like data sources specific to the backend type, as this is covered in official documentation as a legacy approach
- Style guide enforcement beyond `tofu fmt` could catch patterns like:
  - Inconsistent variable naming conventions
  - Missing descriptions in variable and output blocks
  - Overly complex conditional expressions that could be simplified
  - Resource naming patterns that don't follow recommended practices
  
  And frankly, this could even be implemented instead as an _extension to_ `tofu fmt`

For implementation, we could consider a tiered approach:

1. Core rules (in OpenTofu itself):
   - Language-level patterns (like your examples of `depends_on` redundancy and non-converging functions)
   - Documentation-backed best practices
   - Basic style guide enforcement

2. Provider-supplied rules:
   - Provider-specific validations
   - Service-specific best practices
   - Platform limitations and quotas
   (The benefit here would be that there could be official linting for things like handling iam policy documents and other things)

3. External plugin system:
   - Organization-specific rules
   - Cross-provider interactions
   - Custom style guides
   - Industry-specific compliance rules
   - Module input linting guides (probably unlikely)

This would maintain the flexibility that makes tools like `tflint` valuable while providing a solid foundation of core rules that every OpenTofu user can benefit from without additional tooling.

The key difference from the current state would be that anything documented in OpenTofu's official guides as a best practice or anti-pattern would have corresponding lint rules in core, making it easier for users to follow official guidance automatically, which I, anecdotally have seen be pretty common once people learn about the official guidance around things specifically relating to naming and style.

What do you think about this tiered approach to solving the linting problem? I think this effort could begin in stages of most trivial to least, things like formatting resource names and such might be better served as extensions to `tofu fmt` whereas true _linting_ (or at least the way that I think "true linting" is) is more related to quality and code smell prevention.

**Comment by apparentlymart** [comment]
I think that set of tiers does make sense, indeed.

The provider-supplied rules are challenging because that implies a provider protocol extension, which we don't yet have a good means to achieve given that the plugin protocol is driven by the predecessor project. However, I think it would be great to support if we could find a way to do it, to avoid folks needing to install extra stuff.

The external plugin system is, I think, the biggest design question because we'd need to think about who gets to decide which plugins are enabled and where that decision is recorded. For example, it could be something that's decided by a particular module author as part of the module, or it could be something that's defined in the CLI configuration as a set of rules that should be applied to _all_ modules run under that configuration. Perhaps in practice we'd want to allow both, which would then split tier 3 into two sub-tiers 3a and 3b. :grinning: 

When I pondered this before I thought of it as being an extension of `tofu validate` rather than `tofu fmt`, but if we'd imagine having some of the lint rules be automatically-fixable then `tofu fmt` is the one existing command that is expected to perform rewrites. I think we'd need it to be opt-in though because currently `tofu fmt` is exclusively syntax-based and intentionally doesn't start up any provider plugins so that it can be used trivially in contexts like reusable GitHub Actions.

**Comment by joeaawad** [comment]
As mentioned in https://github.com/orgs/opentofu/discussions/1104#discussioncomment-12316023, I wanted to flag that the lack of a linter for OpenTofu is blocking people from adopting any OpenTofu specific features that have syntax that violates the capabilities of tflint. While we are able to migrate to OpenTofu 1.10.6 and use most of the OpenTofu features, we have been prevented from adopting https://github.com/opentofu/opentofu/pull/1718 and similar features because, while that feature would provide significant benefit, it is not worth the increased risk of removing all of our existing linting processes and rules without having an viable replacement. 

Because we have a lot of custom tflint rules, ideally whatever solution is implemented here would be backwards compatible with custom tflint rulesets or at least require minimal changes to existing custom rulesets to migrate the rules to the new solution. Additionally since @dragonfleas mentioned it above, I wanted to call out that one of our existing rules is making sure that module `source` is following our approved patterns and pulling from our shared module source, I see that as probably a fairly common use case where it could make sense to have a built in lint rule that takes some configuration instead of making people write an entirely custom rule for it.

Another thing that I want to flag as we are talking about potential tflint replacement is https://github.com/terraform-linters/tflint/discussions/1690. I would consider this a high impact new feature if the result of this issue is more involved that a minimal change fork of tflint.

**Comment by cpboyd** [comment]
As another example of tflint limitations with respect to OpenTofu: https://github.com/terraform-linters/tflint-ruleset-terraform/issues/305

The useful `for_each` provider definitions ([introduced in v1.9.0](https://opentofu.org/blog/opentofu-1-9-0/)) are not supported, as Terraform itself does not support this.

**Comment by apparentlymart** [comment]
I think it's important to distinguish between "there is no linter for OpenTofu" and "there is no _built-in_ linter in OpenTofu".

Although this issue originally proposed two options the early discussion seemed to suggest that it was primarily about having a _built-in_ linter and I think that's what the OpenTofu maintainers are expecting to eventually make a decision on for the purposes of this issue. (Though as of right now we are not yet ready to make that decision, because it's blocked on a broader decision about how we want to handle extensions to the provider plugin protocol.)

When it comes to there being _any linter at all_ for OpenTofu I just want to be clear that the OpenTofu maintainers don't need to arbitrate _that_ decision at all: anyone in the community could implement a separate linter tool, or fork an existing Terraform linter and adapt it to support OpenTofu's language instead of Terraform's. That seems like the most expedient way to make progress on this question in the immediate term, if anyone is particularly in need of such a thing while we're still figuring out how and whether to support a built-in linter with provider plugin support. Retrofitting some additional language features to an existing linting implementation seems like it ought to be considerably less work than building a new tool inside OpenTofu itself.

**Comment by joeaawad** [comment]
Agreed with @apparentlymart's point that there are two separate ways to address the root cause of this issue with significant differences in complexity. Is there a separate issue to propose that until a first party linter exists, the OpenTofu maintainers should be the group that forks the existing `tflint` and modifies it to be fully compatible with OpenTofu? I know that you all don't need to be the arbiter and anyone in the community could do so, but I see a lot of very strong reasons that it would be best to have a single group governing both OpenTofu and the linter even if those are separate tools. In my opinion, this effort would be the most valuable usage of capacity that the OpenTofu team could do in Q1 of 2026, particularly in light of it being Top 4 on the 👍 list. I just want to make sure that we aren't letting perfect be the enemy of the good because there's only a single issue representing both solutions.

**Comment by apparentlymart** [comment]
There is only a limited amount of maintainer capacity available for the main OpenTofu project, so this one small group of maintainers cannot possibly be responsible for every tool the community might need, and nobody on the current maintainer team has any prior experience working on `tflint` so I don't personally see much advantage to having our existing maintainers _in particular_ work on it vs. anyone else in the community taking it on.

This issue being at number four on the voting list means that there are at least three other significant projects to be done (at least in _some_ sense) before it could become the highest priority for the maintainers -- and we also have to make room for general maintenance work and for supporting the several other forks we've already taken on -- whereas this issue is a great example of something that could happen completely independently of that work if someone else felt it valuable enough to spend their time on it. That would be the most effective way to make progress on this issue in the short term.

If we do spend maintainer time on this then I expect it would be in the form of figuring out how to integrate linting into OpenTofu CLI itself rather than maintaining another separate codebase, because that'd help us to amortize some of the maintenance costs by building on stuff we already have here, since each new codebase we take on has a non-trivial amount of additional overhead in stuff like maintaining release pipelines, reacting to security advisories, etc. But of course we still need to figure out the technical feasibility of that before we can commit to doing it, and that is what I was considering this issue to represent.

If someone needs linting available _right now_ then I do think the best way for them to get it would be to start their own fork of `tflint`. Perhaps in the short term we could use this issue to facilitate community members forming a small team to start and maintain such a fork. Would anyone who is following this issue be interested in working on that?

**Comment by joeaawad** [comment]
Totally understand the concern about limited capacity and lack of familiarity. However, I would like to point out a few things.

First, if the tofulint maintainers are not also the OpenTofu maintainers, that means that the tofulint maintainers will constantly have to play catch up when OpenTofu releases changes that impact the linter. It would be a much better experience for the entire community if OpenTofu and the linter were aligned so that changes to OpenTofu and the linter would get released simultaneously. 

Second, linting is an extremely common capability for any language and if linting capability is not built into the language itself, a linting tool is almost always one of the first tools created. OpenTofu has been lucky that the lack of its own linting tool has not yet been too much of a limiting factor in its growth because most people can use tflint, particularly in the early days. However, this was a significant con that delayed my own company's choice to adopt OpenTofu and I would hazard a guess based on the number of 👍s that this it a fairly common cause of hesitancy across the industry since many companies have policies that all code must be linted. 

Third, if tflint is forked and updated by a random person/group, that can impact the ability of some companies to adopt it because of their security/compliance policies around adopting open source tooling that is not maintained/sponsored by a known/blessed entity. However, if it is owned by the OpenTofu maintainers, those concerns would no longer exist for anyone that is able to use OpenTofu. This would not prevent the community from helping with maintaining the tool, but it would resolve the security/compliance concerns about using it.

Fourth, the lack of a linter is blocking a significant portion of the OpenTofu user base from adopting some of the genuinely great OpenTofu features that you all have rolled out over the past year or so because they are incompatible with tflint and there is no available alternative for OpenTofu right now. While that is currently a viable option, it is getting increasingly less viable as more of these types of features get added and that is also reducing the utility that the OpenTofu team is providing when you add these features because users cannot adopt them. 

Now in terms of how we get to the existence of an OpenTofu linter, I am 100% onboard with the OpenTofu team deciding that this is a Q1 2026 priority and you all want to start the work on planning/building a linter into OpenTofu directly even if that work will take a while. I'm also completely onboard with the OpenTofu team deciding that building linting into OpenTofu directly is too much work for 2026 (or not desired over the long term) and instead deciding to fork and update tflint. And obviously if the OpenTofu team wants to declare linting out of scope for 2026, that's also completely within your rights to do. My ask would be that we get a decision on which of those 3 routes the OpenTofu team would like to take for 2026 within the next few months, and if the OpenTofu team is taking one of the first two routes, a timeline for it to occur so that the rest of the community can plan and respond accordingly.

**Comment by apparentlymart** [comment]
All of the arguments in the previous comment seems like good justification for investing in built-in linting in OpenTofu, and personally I'm hopeful that we'll be able to do that before too long (though I'm certainly not the sole decider) because I agree with all of your arguments that the core lint rules are effectively "part of the language" and should therefore be maintained as a core part of it so that new language features can arrive with appropriate lint rules already implemented[^1].

I was only suggesting a _temporary_ community fork of `tflint` as an interim workaround if you have a need for this that's so urgent that you cannot wait for us to design and implement such a built-in linter. In that I am making a possibly-incorrect assumption that it would be relatively easy for a third-party to patch `tflint` _at least_ so it doesn't treat OpenTofu-specific features as errors, even if it can't generate lint warnings about those features yet. (Caveat that I have not tried this at all, because I'm not familiar with the `tflint` codebase and I also need to be very cautious about reading it since I'm under the impression that current releases contain non-open-source-licensed Terraform code.)

In case it helps folks decide whether making such an investment is worth it, vs. just waiting for a decision on built-in linting, I expect that the earliest possible time we'd be starting to design and implement anything for this is in the v1.13 development period, which is likely to cover the second half of 2026. I'm not promising this will _definitely_ be in that release because we typically only plan one release period at a time, but I _do_ know that all of the maintainers have a pretty full set of planned work for the v1.12 development period (first half of 2026) so we're unlikely to make significant progress on this issue during that time. (Unless the 👍 voting shifts considerably in the next six months, this issue will presumably be closer to the top of the voting list by the time we're planning for v1.13.)

[^1]: We actually already have at least a few warnings that are _effectively_ lint rules -- e.g. the warnings about risky use of `for_each` in `provider` blocks -- but they are not implemented as such because that's not a general concept we have in the codebase yet. I'd love to be able to have more of those, but as long as they are involuntary warnings returned as part of `tofu plan`, rather than opt-in optional checks accessed through a separate command, we can only feasibly offer warnings for situations that have a _very_ low false-positive rate.

**Comment by tmccombs** [comment]
I'll also point out that having it built in has the advantage, at least today, of being able to re-use code from opentofu itself, instead of having to re-implement that for the linter, or do something similar to what tflint does with terraform, and fork opentofu, and patch it to use it as a library for the linter, and then have to stay on top of changes to the upstream project.

*However,* that doesn't necessarily have to be the case. If opentofu exposed a public library that could be used for parsing files, locating modules, loading provider plugins, etc. that would make it significantly for a linter or other external tools to re-use functionality that already exists in opentofu. Perhaps opentofu-schema, or a more complete library built on that  could meet that need?  It seems like that would also be useful for tofu-ls, and perhaps other tools. 

Also, if a built in linter is created, I would very much like to be extensible. It would be especially nice if it could re-use existing tflint plugins.

> and comment on which proposal you would like to support.

I think I would slightly prefer the first proposal (`tofu lint`), but I would be happy with either.

**Comment by gsoeldner** [comment]
Picking up the "fork TFLint into a dedicated OpenTofu linter" idea from this thread — we've made a start and wanted to share it early for feedback and contributors, not as a finished tool.

Repo: https://github.com/SoeldnerConsult/tofulint

Current state: TFLint core forked with OpenTofu internals, basic TFLint functionality working, plus native `.tofu` / `.tofulint.hcl` parsing and pre-bundled Terraform, AWS and GCP rulesets.

It's still experimental. There's a known plugin-init bug (malformed GitHub releases URL → `http: no Host in request URL`) that currently forces manual plugin installation, and it's Linux-only for now.

The part we care about most is the OpenTofu-only linting TFLint won't do — `encryption` blocks first, then provider-defined functions, `removed` blocks and loopable `import` blocks.

Feedback and PRs welcome, especially on the init bug.

**Comment by nikolay** [comment]
@gsoeldner That fork seems dead on arrival.
